pkg/kubectl/cmd/portforward.go

```diff
@@ -131,6 +134,38 @@ func (f *defaultPortForwarder) ForwardPorts(method string, url *url.URL, opts Po
 	return fw.ForwardPorts()
 }
 
+// Translates service port to target port
+// It rewrites ports as needed if the Service port declares targetPort.
+// It returns an error when a named targetPort can't find a match in the pod, or the Service did not declare
+// the port.
+func translateServicePortToTargetPort(ports []string, svc api.Service, pod api.Pod) ([]string, error) {
+	var translated []string
+	for _, port := range ports {
+		// port is in the form of [LOCAL PORT]:REMOTE PORT
+		parts := strings.Split(port, ":")
+		input := parts[0]
+		if len(parts) == 2 {
+			input = parts[1]
+		}
+		portnum, err := strconv.Atoi(input)
+		if err != nil {
+			return ports, err
+		}
+		containerPort, err := util.LookupContainerPortNumberByServicePort(svc, pod, int32(portnum))
+		if err != nil {
+			// can't resolve a named port, or Service did not declare this port, return an error
+			return nil, err
+		} else {
+			if int32(portnum) != containerPort {
```

Comment:
Just realized that if the user does `:443` and the service port 443 has a target port of 8443, the user will still expect 443 locally, so we should translate that to `443:8443`

Reply:
The usage for port-forward is to generate a random port if the local port (the part of string before :) is omitted, so passing :443 means to attach remote port 443 (or what it gets translated into) to a randomly picked local port.  I can add the following test case if that helps people understand the behavior better.

```
                {
                        name: "test success (local port omitted)",
                        svc: api.Service{
                                Spec: api.ServiceSpec{
                                        Ports: []api.ServicePort{
                                                {
                                                        Port: 443,
                                                        TargetPort: intstr.FromString("https"),
                                                },
                                        },
                                },
                        },
                        pod: api.Pod{
                                Spec: api.PodSpec{
                                        Containers: []api.Container{
                                                {
                                                        Ports: []api.ContainerPort{
                                                                {
                                                                        Name:          "https",
                                                                        ContainerPort: int32(8443)},
                                                        },
                                                },
                                        },
                                },
                        },
                        ports:      []string{":443"},
                        translated: []string{":8443"},
                        err:        false,
                },
```

Reply:
Oh, you're right… I misread the examples